src/main/java/com/google/devtools/build/lib/worker/WorkRequestHandler.java

```diff
@@ -496,4 +508,70 @@ private void maybePerformGc() {
       }
     }
   }
+
+  /**
+   * Wrapper class around the standard {@link System} streams that redirects {@link System#out} to
+   * {@link System#err}.
+   *
+   * <p>This is most useful when integrating JVM tools that write exceptions and logs directly to
+   * {@link System#out} which corrupt the persistent worker.
+   *
+   * <p>WorkerIO implements {@link AutoCloseable} and will swap the original streams back into
+   * {@link System} once close has been called.
+   */
+  public static class WorkerIO implements AutoCloseable {
+
+    private final PrintStream originalOutputStream;
+    private final PrintStream originalErrorStream;
+    private final AutoCloseable restore;
+
+    @VisibleForTesting
+    WorkerIO(
+        PrintStream originalOutputStream, PrintStream originalErrorStream, AutoCloseable restore) {
+      this.originalOutputStream = originalOutputStream;
+      this.originalErrorStream = originalErrorStream;
+      this.restore = restore;
+    }
+
+    /**
+     * Creates a new {@link WorkerIO} instance that allows {@link WorkRequestHandler} to safely
+     * redirect {@link System#out} to {@link System#err}.
+     *
+     * <p>{@link WorkerIO#close} must be called to swap the original streams back in.
+     */
+    public static WorkerIO redirectSystemStreams() throws IOException {
+      // Save the original streams
+      PrintStream originalOutputStream = System.out;
+      PrintStream originalErrorStream = System.err;
+
+      // Redirect System.out to System.err
+      System.setOut(originalErrorStream);
+      System.setErr(originalErrorStream);
```

Comment:
This looks a lot simpler. I don't see any reason to do anything to System.err, though.

Reply:
Updated